balances/src/lib.rs

```diff
@@ -251,21 +250,21 @@ impl<T: Config> Pallet<T> {
 		demurrage: BalanceType,
 	) -> BalanceEntry<T::BlockNumber> {
 		let current_block = frame_system::Pallet::<T>::block_number();
-		let elapsed_time_block_number = current_block - entry.last_update;
-		let elapsed_time_u32: u32 = elapsed_time_block_number
-			.try_into()
-			.ok()
-			.expect("blockchain will not exceed 2^32 blocks; qed");
-		let elapsed_time = BalanceType::from_num(elapsed_time_u32);
-		let exponent: BalanceType = -demurrage * elapsed_time;
-		let exp_result: BalanceType = exp(exponent).unwrap();
-		//.expect("demurrage should never overflow");
-		BalanceEntry {
-			principal: entry
-				.principal
-				.checked_mul(exp_result)
-				.expect("demurrage should never overflow"),
-			last_update: current_block,
+
+		match entry.apply_demurrage(demurrage, current_block) {
+			Ok(updated_entry) => updated_entry,
+			Err(e) => {
+				// This should never happen in production! The `apply_demurrage` method is designed
+				// to only return an error when some blockchain invariants no longer hold like:
+				// * last_update < current_block
+				// * demurrage < 0
+				// * blocknumber > 2^32
+				//
+				// Still if such an invariant is violated, what should we do here? Shall we use an
+				// expect after all?
+				debug!("Error when applying demurrage: {:?}", e);
+				Default::default()
+			},
```

Comment:
Here, I honestly don't know what should be the way to handle it. As the function is used inside the fungibles' implementation, we can't just forward an error, so we either panic here or we swallow the error and return something.

Reply:
I added very explicit unreachable for all the errors, which make it clear at pallet level that it should never panic.

Reply:
Changed the demurrage function itself to assume sensible defaults in case an overflow error happens.